express-error-handling/server.ts

```diff
@@ -0,0 +1,134 @@
+/* eslint-disable @typescript-eslint/no-unused-vars -- REMOVE ME */
+import { readFile, writeFile } from 'node:fs/promises';
+import express from 'express';
+import { ClientError } from './client-error.js';
+import { errorMiddleware } from './error-middleware.js';
+
+type Note = {
+  id: number;
+  content: string;
+};
+type Data = {
+  nextId: number;
+  notes: Record<string, Note>;
+};
+
+const app = express();
+app.use(express.json());
+
+async function readData(): Promise<Data> {
+  const data = await readFile('./data.json');
+  return JSON.parse(data.toString());
+}
+
+async function writeData(data: Data): Promise<void> {
+  await writeFile('./data.json', JSON.stringify(data, null, 2));
+}
+
+app.get('/api/notes', async (req, res, next) => {
+  try {
+    const data = await readData();
+    const notes: Note[] = [];
+    for (const id in data.notes) {
+      notes.push(data.notes[id]);
+    }
+    res.json(notes);
+  } catch (err) {
+    console.error(err);
+    res.status(500).json({ error: 'an unexpected error occurred' });
+    next(err);
+  }
+});
+
+app.get('/api/notes/:id', async (req, res, next) => {
+  try {
+    const data = await readData();
+    const id = Number(req.params.id);
+    if (Number.isNaN(id) || !Number.isInteger(id) || id < 1) {
+      res.status(400).json({ error: 'id must be a positive integer' });
```

Comment:
When you throw an error, execution jumps to the catch, and the error middleware handles the response. You do not want to send a response here, you just want to throw the error.